pkg/core/edge_kb.go

```diff
@@ -0,0 +1,296 @@
+	EdgeDetails struct {
+		// ExpansionFunc is a function used to create the To and From resource and necessary intermediate resources, if any do not exist, to ensure the nodes are in place for correct functionality.
+		ExpansionFunc EdgeExpander
+		// Configure is a function used to configure the To and From resources and necessary dependent resources, to ensure the nodes will guarantee correct functionality.
+		Configure EdgeConfigurer
+		// ValidDestinations is a list of end destinations the edge supports. This field is used within determining the path (edge expansion) between two resources.
+		ValidDestinations []reflect.Type
+	}
+
+	// EdgeKB is a map (knowledge base) of edges and their respective details used to configure ResourceGraphs
+	EdgeKB map[Edge]EdgeDetails
+
+	// EdgeExpander is a function used to create the To and From resource and necessary intermediate resources, if any do not exist, to ensure the nodes are in place for correct functionality.
+	EdgeExpander func(from, to Resource, dag *ResourceGraph, data EdgeData) error
+	// EdgeConfigurer is a function used to configure the To and From resources and necessary dependent resources, to ensure the nodes will guarantee correct functionality.
+	EdgeConfigurer func(from, to Resource, dag *ResourceGraph, data EdgeData) error
+
+	// EdgeConstraint is an object defined on EdgeData which can influence the path picked when expansion occurs.
+	EdgeConstraint struct {
+		// NodeMustExist specifies a list of resources which must exist in the path when edge expansion occurs. The resources type will be correlated to the types in the generated paths
+		NodeMustExist []Resource
+		// NodeMustNotExist specifies a list of resources which must not exist when edge expansion occurs. The resources type will be correlated to the types in the generated paths
+		NodeMustNotExist []Resource
+	}
+
+	// EdgeData is an object attached to edges in the ResourceGraph to help the knowledge base understand context when performing expansion and configuration tasks
+	EdgeData struct {
+		// AppName refers to the application name of the global ResourceGraph
+		AppName string
+		// EnvironmentVaribles specify and environment variables which will need to be configured during the edge expansion process
+		EnvironmentVariables []EnvironmentVariable
+		// Constraint refers to the EdgeConstraints defined during the edge expansion
+		Constraint EdgeConstraint
+		// Source refers to the initial source node when edge expansion is called
+		Source Resource
+		// Destination refers to the initial target node when edge expansion is called
+		Destination Resource
+	}
+)
+
+// GetEdgeDetails takes in a source and target to retrieve the edge details for the given key. Will return nil if no edge exists for the given source and target
+func GetEdgeDetails(kb EdgeKB, source reflect.Type, target reflect.Type) EdgeDetails {
```

Comment:
Is there a reason this function and the related functions below don't have an `EdgeKB` receiver?

Reply:
nope thats something i can change and probably is a good idea as we expand what our knowledge base is